src/layout/mod.rs

```diff
@@ -464,6 +478,56 @@ impl<W: LayoutElement> Layout<W> {
         }
     }
 
+    /// Adds a new window to the layout on a specific output & workspace.
+    pub fn add_window_on_workspace(
+        &mut self,
+        output: &Output,
+        workspace_name: &str,
+        window: W,
+        width: Option<ColumnWidth>,
+        is_full_width: bool,
+    ) {
+        let mut width = width.unwrap_or_else(|| ColumnWidth::Fixed(window.size().w));
+        if let ColumnWidth::Fixed(w) = &mut width {
+            let rules = window.rules();
+            let border_config = rules.border.resolve_against(self.options.border);
+            if !border_config.off {
+                *w += border_config.width as i32 * 2;
+            }
+        }
+
+        let MonitorSet::Normal {
+            monitors,
+            active_monitor_idx,
+            ..
+        } = &mut self.monitor_set
+        else {
+            panic!()
+        };
+
+        let (mon_idx, mon) = monitors
+            .iter_mut()
+            .enumerate()
+            .find(|(_, mon)| mon.output == *output)
+            .unwrap();
+
+        let ws_idx = &mon
+            .find_workspace_index(workspace_name)
+            .unwrap_or(mon.active_workspace_idx);
```

Comment:
I believe this family of functions is supposed to panic when supplied invalid arguments, so this can just `.unwrap()`.

Reply:
I *think* my thinking here was that if a named workspace disappears between initially resolving it in the window rules, and mapping the window, then this would gracefully fall back to using the current workspace instead.

I'm not sure such a thing can happen now. Or ever. I'll change it to an `unwrap()`, because the rest of the stuff panics there on error anyway. If it ever becomes a problem, that can be fixed *then*.

Reply:
Hmm, that's actually a good point. But I think you could check it outside this function, in the commit handler. It has a better idea of where to fall back in that case.

Reply:
Fwiw, any amount of time can pass between the initial configure and the window finally deciding to commit its first buffer. Usually windows commit more or less right away (they intend to show up on screen after all) but it is not a given (and even that tiny amount of time can be enough), so yeah, assume anything can change in between the initial configure and the window being mapped.